### PR TITLE
APPSI: Auto-Persistent Pyomo Solver Interfaces

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
This PR introduces a new package in contrib called APPSI. APPSI is a set of persistent solver interfaces for efficiently resolving models with small changes. 

This PR is not ready to merge. However, I wanted to create the PR so those interested could start taking a look.

Installation requires compiling a small amount of c++ code using pybind11. All you need to do is install pybind11 and run `python build.py build` from the `appsi` directory.

I am also going to use this PR for a checklist.

- [ ] Build and distribute
  - [x] Integrate into pyomo build-extenstions
  - [ ] add to cythonized wheel builder
- [x] Improve testing
  - [x] Test examples
  - [x] Test more nonlinear functions in the Ipopt interface
  - [x] Test Gurobi callbacks
  - [x] Test linear expressions
- [x] Generate documentation
  - [x] Basic usage
  - [x] Handling model updates
  - [x] Installation
- [ ] Test and document performance
  - [ ] Test overhead of caching things for later updates (i.e., can all solver interfaces be persistent?)
  - [ ] Basic performance tests for cut generation, parameter changes, etc.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by michaelbynum** [comment]
Temporarily closing this.

**Review by carldlaird — APPROVED** [review]
looks like a reasonable contrib package

File: .github/workflows/test_branches.yml [hunk 1]
```diff
@@ -231,6 +236,7 @@ jobs:
         python -m pip install --cache-dir cache/pip --upgrade pip
         pip install --cache-dir cache/pip ${PYTHON_CORE_PKGS}
         pip install --cache-dir cache/pip ${PYTHON_PACKAGES}
+        pip install --cache-dir cache/pip pybind11
```
Comment (jsiirola) [inline .github/workflows/test_branches.yml:239]: `pybind11` should be added with parameterized to `PYTHON_REQUIRED_PKGS`

File: .github/workflows/test_branches.yml [hunk 2]
```diff
@@ -261,7 +267,10 @@ jobs:
         conda list --show-channel-urls
         conda install -q -y -c conda-forge ${PYTHON_CORE_PKGS}
         conda install -q -y -c conda-forge ${PYTHON_PACKAGES}
+        conda install -q -y -c conda-forge pybind11
         if test -z "${{matrix.slim}}"; then
+            conda install -q -y -c gurobi gurobi \
+                || echo "WARNING: Gurobi is not available"
```
Comment (jsiirola) [inline .github/workflows/test_branches.yml:273]: For sanity, can you install things in the same order as with pip?

File: pyomo/common/collections/orderedset.py [hunk 3]
```diff
@@ -80,3 +80,11 @@ def clear(self):
...
         """Remove an element. If not a member, raise a KeyError."""
         del self._dict[val]
+
+    def intersection(self, other):
+        res = OrderedSet([i for i in self if i in other])
+        return res
+
+    def union(self, other):
+        res = OrderedSet(self)
+        res.update(other)
```
Comment (jsiirola) [inline pyomo/common/collections/orderedset.py:90]: Can you ass a test covering this new method?

**Review by jsiirola — COMMENTED** [review]
A couple quick changes to the test infrastructure, and I think this will be fine for an initial merge.

**Pull request merged**